### ISSUE TITLE
CombatTracker invocation of `initialize` in constructor prevents child class override of `initialize` w/ private variables or methods.

### ISSUE THREAD

**Issue by typhonrt** [foundryvtt/foundryvtt#11976]
### What happened?

This is a particular edge case, but one worth fixing as the fix is easy. The `CombatTracker` class invokes `initialize` as the last statement of the constructor. If a child class overrides `initialize` and uses private variables / other private methods this causes an issue due to that the child class constructor has yet to be executed. 

Now granted the `CONFIG.ui` class entries are automatically instantiated blindly in `Game.initializeUI` via:

```js
    // Initialize all singleton applications
    for ( let [k, cls] of Object.entries(CONFIG.ui) ) {
      ui[k] = new cls();
    }
```

The fix in this case is use a timeout to defer `initialize` to the next clock tick in the `CombatTracker` constructor. This allows the child class constructor to complete before `initialize` is invoked and allows the ability to override `initialize` in a child class implementation and use private scoped variables / methods.

I haven't checked if this pattern of invoking `initialize` or similar method in a constructor is used in other `CONFIG.ui` default implementations, but worth checking / adding a timeout in most cases if a method is invoked before child class construction is complete.

Change this: 
```js
CombatTracker {
  constructor(options) {
    // Other things above...
    
    this.initialize({render: false});
  }
}
```

To this: 
```js
CombatTracker {
  constructor(options) {
    // Other things above...
    
    // Defer initialization until next clock tick.
    setTimeout(() => this.initialize({render: false}), 0);
  }
}
```

### What ways of accessing Foundry can you encounter this issue in?

- [X] Native App (Electron)
- [X] Chrome
- [X] Firefox
- [X] Safari
- [ ] Other

### Reproduction Steps

- Create a custom CombatTracker extending the default one. 

- Override `initialize`. Use `super.initialize(options)`... 

- Try to use private variables in an overridden `initialize` implementation.

### What core version are you reporting this for?

v12.331

### Bug Checklist

- [X] The issue occurs while all Modules are disabled

**Comment by aaclayton** [comment]
You can't access private variables of the parent class in the subclass. This doesn't have anything to do with construction/initialization order.

**Comment by Ethaks** [comment]
While I do no necessarily agree with the timeout solution, the issue as I understand it is not about using the private fields of Foundry's class, but ones of the extending one:
```js
class Foo {
	#foo = 0;
	constructor() {
		this.init();
	}
	init() {
		console.log("Foo#init", this.#foo);
		this.#foo = this.#foo + 1;
		console.log("Foo#init", this.#foo);
	}
}

class Bar extends Foo {
	#bar = 0;
	init() {
		super.init();
		console.log("Bar#init");
		this.#bar = this.#bar + 1; // This will error
		console.log("Bar#init", this.#bar);
	}
}
```

Above code will error despite `Bar` only accessing a private field of its own making in its own `init`, simply because the call order will result in `init` being run as `Foo` and not an instance of `Bar` yet, with the output being
```
Foo#init 0
Foo#init 1
Bar#init
Uncaught TypeError: can't access private field or method: object is not the right class
```

**Comment by typhonrt** [comment]
Yes Ethaks has the correct understanding. I'm well aware of how scoping works / informed developer, etc. 

[queueMicrotask](https://developer.mozilla.org/en-US/docs/Web/API/Window/queueMicrotask) is the most appropriate solution. I'll be trying that out in ~30 min...

**Comment by typhonrt** [comment]
Yep.. I can confirm that `queueMicrotask` fixes the issue:

```js
CombatTracker {
  constructor(options) {
    // Other things above...
    
    queueMicrotask(() => this.initialize({render: false}));
  }
}
```

----

If you want to see the actual problem here is a simple test case:

```js
class BrokenCombatTracker extends CombatTracker
{
   #bug = true;

   constructor(options)
   {
      super(options);
      console.log(`!!! BrokenCombatTracker - constructor`);
   }

   initialize(options)
   {
      console.log(`!!! BrokenCombatTracker - init - 1`);
      super.initialize(options);

      console.log(`!!! Is there a bug: ${this.#bug}`);

      console.log(`!!! BrokenCombatTracker - init - GOT HERE`);
   }
}

Hooks.once('init', () =>
{
   CONFIG.ui.combat = BrokenCombatTracker;
});
````

**Comment by aaclayton** [comment]
This "solution" would significantly change the expectations of the design however. A queued microtask is effectively an un-awaited async workflow and therefore any preparation of instance properties or other data is unavailable at the moment that the constructor finishes. 

For example:

```js
class Foo {
  constructor() {
    queueMicrotask(this.initialize.bind(this));
  }

  bar;

  initialize() {
    this.bar = 1;
  }
}

const f = new Foo();
console.log(f.bar); // undefined when 1 is expected
```

I think it may be more unambiguous for us to simply remove `ui.combat.initialize()` from the constructor entirely and call it explicitly as a post-constructor operation. Comfortable taking whichever option @Fyorl prefers.

**Comment by typhonrt** [comment]
Possibly create a private `#initializeImpl` that `initialize` invokes. 

```js
class CombatTracker {
   constructor() {
     // Other stuff
     this.#initializeImpl({render: false});
   }
   
   initialize(options) {
      this.#initializeImpl(options)
   }
   
   #initializeImpl({combat=null, render=true}={}) {
      // Where the magic happens that was previously in `initialize`.
   }
}
```

**Comment by Fyorl** [comment]
`CombatTracker#initialize` is now deprecated, and is no longer called in the `constructor`. Changing the viewed combat can now be done as part of a render call, i.e. `ui.combat.render({ combat })`, or by assigning to `ui.combat.viewed` directly if a re-render is not desired.
